src/bin/configurations/mod.rs

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use clap::{parser::ValueSource, ArgMatches};
+use solang::{
+    codegen::{OptimizationLevel, Options},
+    file_resolver::FileResolver,
+    Target,
+};
+
+use std::{collections::HashSet, ffi::OsString, fs, path::PathBuf, process::exit};
+use toml::Table;
+mod unit_tests;
+
+// A macro for error handling.
+macro_rules! match_err {
+    ($func_name: expr) => {
+        match $func_name {
+            Ok(value) => value,
+            Err(err) => {
+                return Err(format!("{err}"));
+            }
+        }
+    };
+}
+
+// A type that defines the package to be compiled: file names, contract names, import maps and import paths.
+type Package = (
+    Vec<OsString>,
+    HashSet<String>,
+    Vec<PathBuf>,
+    Vec<(String, PathBuf)>,
+);
+
+type CompilerOutput = (
+    bool,
+    bool,
+    Option<String>,
+    Option<OsString>,
+    Option<OsString>,
+);
+
+// All compile subcommand configurations
+pub struct Configurations {
+    pub options: Options,
+    pub verbose: bool,
+    pub target: Target,
+    pub filenames: Vec<OsString>,
+    pub contract_names: HashSet<String>,
+    pub imports: FileResolver,
+    pub emit: Option<String>,
+    pub std_json_output: bool,
+    pub output_directory: Option<OsString>,
+    pub output_meta: Option<OsString>,
+}
+
+impl Configurations {
+    /// Parse compile configurations from a toml file.
+    pub fn config_from_toml(path: &PathBuf) -> Result<Self, String> {
```

Comment:
Why do we want to manually parse the TOML file instead of using serde and [the TOML crate](https://crates.io/crates/toml)?